### PR TITLE
Read gtfs routes

### PR DESCRIPTION
**Pull request by datanel**

### CODE REVIEW

**Comment by TeXitoi** [comment]
I'm on it

File: src/gtfs/read.rs [hunk 1]
```diff
@@ -204,7 +208,7 @@ impl<'de> ::serde::Deserialize<'de> for RouteType {
     }
 }
 
-#[derive(Serialize, Deserialize, Debug, Clone)]
+#[derive(Serialize, Deserialize, Debug, Clone, Eq, Hash, PartialEq)]
```
Comment (TeXitoi) [inline src/gtfs/read.rs]: why do you need Eq and Hash?

File: src/gtfs/read.rs [hunk 2]
```diff
@@ -410,24 +455,28 @@ fn get_modes_from_gtfs(
...
+            None
+        } else {
+            Some(r.short_name.to_string())
+        }
+    };
+
+    let line_agency = |r: &Route| match r.agency_id {
+        Some(ref agency_id) => agency_id.to_string(),
+        None => default_agency_id(),
+    };
```
Comment (TeXitoi) [inline src/gtfs/read.rs]:
```rust
let line_agency = |r: &Route| r.agency_id.as_ref().map(|id| id.to_string()).unwrap_or_else(|| default_agency_id());

File: src/gtfs/read.rs [hunk 3]
```diff
@@ -445,24 +494,212 @@ fn get_lines_from_gtfs(gtfs_routes: &[Route], read_mode: &RouteReadType) -> Vec<
...
+                        backward_routes.insert(r2);
+                        if r1.id > r2.id {
+                            routes.remove(r1);
+                            routes.insert(r2);
+                        }
+                    } else {
+                        routes.insert(r2);
+                    }
+                }
+            }
```
Comment (TeXitoi) [inline src/gtfs/read.rs]: I don't understand this code.

File: src/gtfs/read.rs [hunk 4]
```diff
@@ -445,24 +494,212 @@ fn get_lines_from_gtfs(gtfs_routes: &[Route], read_mode: &RouteReadType) -> Vec<
...
+                    object_properties: KeysValues::default(),
+                    comment_links: CommentLinksT::default(),
+                    line_id: r.id.to_string(),
+                    geometry_id: None,
+                    destination_id: None,
+                };
+                routes.push(route);
+                if backward_routes.contains(&r.id) {
+                    let route = objects::Route {
+                        id: r.id.to_string() + "_R",
```
Comment (TeXitoi) [inline src/gtfs/read.rs]: what about clash?

File: src/gtfs/read.rs [hunk 5]
```diff
@@ -445,24 +494,212 @@ fn get_lines_from_gtfs(gtfs_routes: &[Route], read_mode: &RouteReadType) -> Vec<
...
+                    routes.push(route);
+                } else if current_route_directions.is_empty() {
+                    panic!("RouteAsNtmRoute - Coudn't find trips for route_id {}", r.id);
+                } else {
+                    let direction = current_route_directions.iter().next().unwrap();
+                    let direction_name = if *direction == DirectionType::Forward {
+                        "forward"
+                    } else {
+                        "backward"
+                    };
```
Comment (TeXitoi) [inline src/gtfs/read.rs]: better exhastive match or a display implementation or something similar

File: src/gtfs/read.rs [hunk 6]
```diff
@@ -445,24 +494,212 @@ fn get_lines_from_gtfs(gtfs_routes: &[Route], read_mode: &RouteReadType) -> Vec<
...
+                }
+            }
+        }
+        RouteReadType::RouteAsNtmRoute => {
+            let routes_directions: HashSet<(String, DirectionType)> = gtfs_trips
+                .iter()
+                .map(|t| (t.route_id.clone(), t.direction.clone()))
+                .collect();
+
+            let same_lines = get_same_lines(gtfs_routes);
```
Comment (TeXitoi) [inline src/gtfs/read.rs]: I think we can do simpler for this block

File: src/gtfs/read.rs (hunk 1 above)
Comment (datanel) [reply src/gtfs/read.rs]: for HashSet I suppose. It does not compile without

File: src/gtfs/read.rs (hunk 3 above)
Comment (datanel) [reply src/gtfs/read.rs]: suppose you have 3 routes r2, r21 r3 (r1, r2 are the same line)
we'll have 2 lines (r1, r3) (r1 = the smallest id between r1 and r2)

1. I insert r2 
2. r1<r2 => I replace r2 by r1
3. I insert r3
. finally I have my 2 lines r1, r3

poor algo but if you have better..

File: src/gtfs/read.rs (hunk 4 above)
Comment (datanel) [reply src/gtfs/read.rs]: clash ? the rock band? I don't understand

File: src/gtfs/read.rs [hunk 7]
```diff
@@ -448,61 +472,147 @@ fn get_modes_from_gtfs(
...
 }
 
-fn get_lines_from_gtfs(gtfs_routes: &[Route], read_mode: &RouteReadType) -> Vec<objects::Line> {
+fn get_route_with_smallest_name<'a>(routes: &'a [&Route]) -> &'a Route {
+    routes
+        .into_iter()
+        .fold(None, |min, x| match min {
+            None => Some(x),
+            Some(y) => Some(if x.id < y.id { x } else { y }),
+        })
```
Comment (TeXitoi) [inline src/gtfs/read.rs]: https://doc.rust-lang.org/std/iter/trait.Iterator.html#method.min_by_key

File: src/gtfs/read.rs [hunk 8]
```diff
@@ -448,61 +472,147 @@ fn get_modes_from_gtfs(
...
+            for t in gtfs_trips.iter().filter(|t| t.route_id == r.id) {
+                route_directions.insert(&t.direction);
+            }
+            if route_directions.is_empty() {
+                ensure!(
+                    !route_directions.is_empty(),
+                    "Coudn't find trips for route_id {}",
+                    r.id,
+                );
+            }
```
Comment (TeXitoi) [inline src/gtfs/read.rs]: @prhod we fail if a route has no trip?

Comment (prhod) [reply src/gtfs/read.rs]: If possible, please don't panic but add a warning  log.
Can you avoid creating empty Routes (and therefore empty lines) ?

File: src/gtfs/read.rs [hunk 9]
```diff
@@ -513,26 +507,30 @@ fn make_lines(map_line_routes: &MapLineRoutes) -> Vec<objects::Line> {
...
-            color: r.color.clone(),
-            text_color: r.text_color.clone(),
-            sort_order: r.sort_order,
-            network_id: line_agency(r),
-            commercial_mode_id: r.route_type.to_gtfs_value(),
-            geometry_id: None,
-            opening_time: None,
-            closing_time: None,
-        });
+        let route_has_trips = gtfs_trips.iter().find(|t| t.route_id == r.id);
```
Comment (TeXitoi) [inline src/gtfs/read.rs]: I'd put the is_some on this line.

File: src/gtfs/read.rs [hunk 10]
```diff
@@ -473,13 +473,7 @@ fn get_modes_from_gtfs(
...
 
 fn get_route_with_smallest_name<'a>(routes: &'a [&Route]) -> &'a Route {
-    routes
-        .into_iter()
-        .fold(None, |min, x| match min {
-            None => Some(x),
-            Some(y) => Some(if x.id < y.id { x } else { y }),
-        })
-        .unwrap()
+    routes.iter().min_by_key(|&r| r.id.clone()).unwrap()
```
Comment (TeXitoi) [inline src/gtfs/read.rs]: without clone

**Pull request merged**